packages/common/infra/src/workspace/engine/doc/README.md

```diff
@@ -0,0 +1,127 @@
+# DocEngine
+
+The synchronization algorithm for yjs docs.
+
+```
+ ┌─────────┐  ┌───────────┐  ┌────────┐
+ │ Storage ◄──┤ DocEngine ├──► Server │
+ └─────────┘  └───────────┘  └────────┘
+```
+
+# Core Components
+
+## DocStorage
+
+```ts
+export interface DocStorage {
+  eventBus: DocEventBus;
+  doc: ByteKV;
+  syncMetadata: ByteKV;
+  serverClock: ByteKV;
+}
+```
+
+Represents the local storage used, Specific implementations are replaceable, such as `IndexedDBDocStorage` on the `browser` and `SqliteDocStorage` on the `desktop`.
+
+### DocEventBus
+
+Each `DocStorage` contains a `DocEventBus`, which is used to communicate with other engines that share the same storage.
+
+With `DocEventBus` we can sync updates between engines without connecting to the server.
+
+For example, on the `browser`, we have multiple tabs, all tabs share the same `IndexedDBDocStorage`, so we use `BroadcastChannel` to implement `DocEventBus`, which allows us to broadcast events to all tabs.
```

Comment:
will the two tabs get the same update at the same time from server if others pushed one?

Reply:
technically, we can allow receive updates from the server and `BroadcastChannel` at the same time, and they can be merged.

however, if we accept updates through `BroadcastChannel`, we better elect a tab to stay connected to the server, and all other tabs should be disconnected from the server and idb to avoid wasted bandwidth and multiple writes (idb)

Reply:
after checking the code, currently no logic to disconnect ws when bc is connected, so updates will be received from the server twice and write to idb twice

Reply:
All tabs have their own ws links, 

Based on the current design, we will not choose one tab as the leader. All tabs are equal and can be closed at any time.

Reply:
If in the future we have a `service worker` we can just do the ws connection in the `service worker`.